x-pack/plugins/observability_solution/apm/public/components/app/transaction_details/waterfall_with_summary/waterfall_container/waterfall/index.tsx

```diff
@@ -59,35 +63,36 @@ function getWaterfallMaxLevel(waterfall: IWaterfall) {
   if (!entryId) {
     return 0;
   }
+
   let maxLevel = 1;
-  function countLevels(id: string, currentLevel: number) {
+  const visited = new Set<string>();
+  const queue: Array<{ id: string; level: number }> = [{ id: entryId, level: 1 }];
+
+  while (queue.length > 0) {
+    const { id, level } = queue.shift()!;
     const children = waterfall.childrenByParentId[id] || [];
-    if (children.length) {
-      children.forEach((child) => {
-        // Skip processing when a child node has the same ID as its parent
-        // to prevent infinite loop
-        if (child.id !== id) {
-          countLevels(child.id, currentLevel + 1);
-        }
-      });
-    } else {
-      if (maxLevel < currentLevel) {
-        maxLevel = currentLevel;
+
+    maxLevel = Math.max(maxLevel, level);
+    visited.add(id);
+
+    children.forEach((child) => {
+      if (child.id !== id && !visited.has(child.id)) {
+        queue.push({ id: child.id, level: level + 1 });
+        visited.add(child.id);
       }
-    }
+    });
   }
 
-  countLevels(entryId, 1);
   return maxLevel;
 }
+
 // level starts with 0
```

Comment:
this comment is a little confusing to me given that we initialize `maxLevel` with 1 on line 67. Also maybe we can use screaming caps here to make clear that it's really a constant. e.g. `MAX_LEVEL_OPEN`.

Reply:
the variable name is also confusing. I'll change that.